### PR TITLE
Storage

### PR DESCRIPTION
**Pull request by jreidinger**

### CODE REVIEW

**Comment by mvidner** [comment]
Today's yast-storage commit adds a function which is not stubbed yet: 

```
src/modules/Storage.ycp:6368  [Parser] Unknown identifier 'LibStorage::getPresentDisks'
```

**Comment by jreidinger** [comment]
Updated

**Comment by mvidner** [comment]
OK, there are some "dynamic constant assignment" errors in modules, but this is partial progress which will unblock some dependencies.

File: patches/storage.patch [hunk 1]
```diff
@@ -0,0 +1,1479 @@
...
++    import "StorageSettings";
+ 
+     include "partitioning/ep-lib.ycp";
+     include "partitioning/custom_part_lib.ycp";
+@@ -724,8 +734,21 @@ because the filesystem seems to be inconsistent.
+ 	switch (data["type"]:`unknown)
+ 	{
+ 	    case `primary:
++        {
++                // no falldown in ruby
```
Comment (mvidner) [inline patches/storage.patch:169]: I think this is wrong to duplicate code in this way. @dmajda: IMHO y2r should be improved to generate `case :primary, :logical` for this.

File: patches/storage.patch [hunk 2]
```diff
@@ -0,0 +1,1479 @@
...
+ 			{
+ 			    case `primary:
++            {
++				integer num_cyl = tointeger(1024.0 * size_k / cyl_size + 0.5);
++				num_cyl = Integer::Clamp(num_cyl, 1, data["region", 1]:0 + free_cyl_after);
++				data["region", 1] = num_cyl;
++            }
++				break;
++
+ 			    case `logical:
```
Comment (mvidner) [inline patches/storage.patch:254]: again a "multiwhen"

File: patches/storage.patch [hunk 3]
```diff
@@ -0,0 +1,1479 @@
...
+ 	    switch (widget)
+ 	    {
+ 		case `pool:
++                    //no falldown in ruby
++		    UI::ChangeWidget(`id(`used_pool), `Enabled, (widget==`thin));
++		    break;
+ 		case `thin:
++                    //no falldown in ruby
++		    UI::ChangeWidget(`id(`used_pool), `Enabled, (widget==`thin));
++		    break;
```
Comment (mvidner) [inline patches/storage.patch:462]: multiwhen

File: patches/storage.patch [hunk 4]
```diff
@@ -0,0 +1,1479 @@
...
++        // no falldown in ruby
++		    huhu(disk, `hd, $[ `create : CreateHdPartitionPanel, `handle : HandleHdPartitionPanel ],
++			 $[ `create : CreateHdDiskPanel, `handle : HandleHdDiskPanel ]);
++		    break;
+ 		case `CT_DMMULTIPATH:
++        // no falldown in ruby
++		    huhu(disk, `hd, $[ `create : CreateHdPartitionPanel, `handle : HandleHdPartitionPanel ],
++			 $[ `create : CreateHdDiskPanel, `handle : HandleHdDiskPanel ]);
++		    break;
+ 		case `CT_MDPART:
```
Comment (mvidner) [inline patches/storage.patch:519]: multiwhen

File: patches/storage.patch [hunk 5]
```diff
@@ -0,0 +1,1479 @@
...
++		break;
++
+ 	    case `CT_DMRAID:
++        // no falldown in ruby
++		if (part == nil)
++		    EpContextMenuHdDisk(device);
++		else
++		    EpContextMenuHdPartition(device);
++		break;
++
```
Comment (mvidner) [inline patches/storage.patch:548]: multiwhen

File: patches/storage.patch [hunk 6]
```diff
@@ -0,0 +1,1479 @@
...
+ 	    case `CT_DMMULTIPATH:
+ 		if (part == nil)
+ 		    EpContextMenuHdDisk(device);
+@@ -463,16 +496,23 @@ Really exit?"), deletechars(label, "&"));
+ 
+ 	    switch (widget)
+ 	    {
+-		case `back:
+-		case `abort: {
```
Comment (mvidner) [inline patches/storage.patch:557]: multiwhen

File: patches/storage.patch [hunk 7]
```diff
@@ -0,0 +1,1479 @@
...
+ 	until (widget == `back || widget == `abort || widget == `next);
+@@ -513,6 +553,11 @@ Really exit?"), deletechars(label, "&"));
+ 	switch (widget)
+ 	{
+ 	    case `back:
++        // no falldown in ruby
++		if (Storage::GetPartMode() == "NORMAL")
++		    Storage::SetPartMode("CUSTOM");
++
++		break;
```
Comment (mvidner) [inline patches/storage.patch:608]: multiwhen

File: patches/storage.patch [hunk 8]
```diff
@@ -0,0 +1,1479 @@
...
+ 		case `raid5:
++                    //no falldown in ruby
++		    raid_type = substring(tostring((symbol) UI::QueryWidget(`id(`raid_type), `Value)), 1);
++		    DevicesSelectionBox::UpdateSelectedSize();
++		break;
+ 		case `raid6:
++                    //no falldown in ruby
++		    raid_type = substring(tostring((symbol) UI::QueryWidget(`id(`raid_type), `Value)), 1);
++		    DevicesSelectionBox::UpdateSelectedSize();
++		break;
```
Comment (mvidner) [inline patches/storage.patch:678]: multiwhen

File: patches/storage.patch [hunk 9]
```diff
@@ -0,0 +1,1479 @@
...
+ 	{
+-	    case `unselected:
++	    case `unselected: {
++    // no fall thrue
++		list<any> tmp1 = (list<any>) UI::QueryWidget(`id(`unselected), `SelectedItems);
++		selected = flatten([ selected, tmp1 ]);
++		sel = (list<string>) UI::QueryWidget(`id(`selected), `SelectedItems);
++                y2milestone( "selected:%1", selected );
++	    } break;
++
```
Comment (mvidner) [inline patches/storage.patch:778]: multiwhen

File: patches/storage.patch [hunk 10]
```diff
@@ -0,0 +1,1479 @@
...
+-	    case `selected:
++	    case `selected: {
++    // no fall thrue
++list<any> tmp1 = (list<any>) UI::QueryWidget(`id(`selected), `SelectedItems);
++		selected = filter(any tmp2, selected, { return !contains(tmp1, tmp2); });
++		sel = (list<string>) UI::QueryWidget(`id(`selected), `SelectedItems);
++                sel = (list<string>)filter(any tmp2, sel, { return !contains(tmp1, tmp2); });
++                y2milestone( "selected:%1", selected );
++	    } break;
++
```
Comment (mvidner) [inline patches/storage.patch:796]: multiwhen

File: patches/storage.patch [hunk 11]
```diff
@@ -0,0 +1,1479 @@
...
++                    integer idx = sm[s]:0;
++                    if( (up && idx>cnt) || (!up && idx<siz-cnt) )
++                        {
++                        selected = list::swap( selected, idx+diff, idx );
++                        }
++                    cnt=cnt+1;
++                    });
++                y2milestone( "change:%1 selected:%2", sel, selected  );
++	    } break;
++
```
Comment (mvidner) [inline patches/storage.patch:830]: multiwhen

File: patches/storage.patch [hunk 12]
```diff
@@ -0,0 +1,1479 @@
...
++    // no fall thrue
++                boolean up = (widget==`top);
++		sel = (list<string>) UI::QueryWidget(`id(`selected), `SelectedItems);
++                selected = filter( string s, (list<string>)selected, ``(!contains( sel, s )));
++                if( up )
++                    selected = merge( sel, selected );
++                else
++                    selected = merge( selected, sel );
++                y2milestone( "change:%1 selected:%2", sel, selected  );
++	    } break;
```
Comment (mvidner) [inline patches/storage.patch:849]: multiwhen

**Comment by mvidner** [comment]
Number of multiwhens: 12.

File: patches/storage.patch (hunk 1 above)
Comment (dmajda) [inline patches/storage.patch:169]: @jreidinger @mvidner I'll make Y2R handle this case.

**Comment by mvidner** [comment]
Review done: the multiwhens are my only objection. @dmajda if you agree to implement them we can merge this.

**Pull request merged**